src/core/expression/qgsexpressionfunction.cpp

```diff
@@ -6730,6 +6750,126 @@ static QVariant executeGeomOverlay( const QVariantList &values, const QgsExpress
 
     if ( ! relationFunction || ( geometry.*relationFunction )( feat2.geometry() ) ) // Calls the method provided as template argument for the function (e.g. QgsGeometry::intersects)
     {
+
+      if ( isIntersectsFunc && ( minOverlap != -1 || minInscribedCircleRadius != -1 ) )
+      {
+        const QgsGeometry intersection { geometry.intersection( feat2.geometry() ) };
+
+        switch ( intersection.type() )
+        {
+          case QgsWkbTypes::GeometryType::PolygonGeometry:
+          {
+            // Check min overlap for intersection (if set)
+            if ( intersection.isMultipart() )
+            {
+              bool testResult { false };
+              for ( auto it = intersection.const_parts_begin(); ! testResult && it != intersection.const_parts_end(); ++it )
+              {
+                const QgsPolygon *geom = qgsgeometry_cast< const QgsPolygon * >( *it );
+                // qDebug() << "Area" << feat2.id() << geom->area();
+                if ( minOverlap != -1 )
+                {
+                  if ( geom->area() >= minOverlap )
+                  {
+                    testResult = true;
+                  }
+                  else
+                  {
+                    continue;
+                  }
+                }
+
+                // Check min inscribed circle radius for intersection (if set)
+                if ( minInscribedCircleRadius != -1 )
+                {
+                  const QgsRectangle bbox = geom->boundingBox();
+                  const double width = bbox.width();
+                  const double height = bbox.height();
+                  const double size = width > height ? width : height;
+                  const double tolerance = size / 1000.0;
```

Comment:
shouldn't the tolerance be dependent on the minInscribedCircleRadius value instead of the geometry's size?

Reply:
I'm not sure: I used the same logic that Postgis implements: https://github.com/postgis/postgis/commit/b5bd6836d4617ebc924e804c42b60de637a94f60#diff-33eafe0a60e374babff748cb4f7897d65e6ea040e2cc3e4602bf2b459adc223bR398

Reply:
That's a different use case -- they are setting the tolerance for the function which returns the maximum inscribed circle. In that case the overall size is a good basis to use. Here we don't care about the **real** maximum inscribed circle, just whether there is one anywhere which exceeds that tolerance size. So it's an inverse relationship here -- the smaller the tolerance, the less strict we need to be in finding the circle. The larger the tolerance the more strict we'll need to be. 

That all said... I can't actually think of a good way to use this inverse relationship to pick a suitable tolerance!

I would say though that the 1000 value is likely excessive and slow. Do you get reasonable results if you drop this to somewhere between 20-100?

Reply:
I don't have access to a real large use case to test with, maybe @andreasneumann can help to check if changing hardcoded tolerance from 1000 to 100 makes a difference in speed and reliability of the function?

There is also a possibility to add another argument to set the tolerance, I was hoping that using the same approach that postgis used could have spared us an extra argument.

Another possible optimization is probably to check the bbox in advance and skip the actual calculation if none of the bbox halved dimensions exceed the min radius.

Reply:
I've reduced tolerance denominator to 100, let's give it some tests.